### PR TITLE
 [QoS]Adding support to read and verify the VOQ counters in LossyQueueTest

### PR DESCRIPTION
**Pull request by ansrajpu-git**
### Description of PR

Summary:
Adding support to read and verify the VOQ counters in LossyQueueTest for the dnx platform.
Below are the changes made to achieve this:
- Collect all the system ports associated with the destination port of the DUT (qos_sai_base.py)
- Get the voq_id for all for the corresponding system ports and read the voq_counters( switch.py)
- Read & verify the voq_counters in Test (sai_qos_tests.py) 
 
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205
- [ ] 202305

#### How did you do it?
Below are the changes made to support voq counters in LossyQueueTest:
- Collect all the system ports associated with the destination port of the DUT (qos_sai_base.py)
- Get the voq_id for all for the corresponding system ports and read the voq_counters( switch.py)
- Read & verify the voq_counters in Test (sai_qos_tests.py)     

#### How did you verify/test it?
Executed the Qos tests

### CODE REVIEW

File: tests/qos/qos_sai_base.py [hunk 1]
```diff
@@ -934,24 +954,43 @@ def dutConfig(
             src_asic = get_src_dst_asic_and_duts['src_asic']
             dst_dut_index = get_src_dst_asic_and_duts['dst_dut_index']
             dst_asic = get_src_dst_asic_and_duts['dst_asic']
+            src_system_port = src_dut.config_facts(host=src_dut.hostname, source='running')[
```
Comment (vmittal-msft) [inline tests/qos/qos_sai_base.py:957]: Can we add vendor/chip specific check as it system_port may not be applicable for all ?

File: tests/qos/qos_sai_base.py [hunk 2]
```diff
@@ -934,24 +954,43 @@ def dutConfig(
...
             testPortIds[src_dut_index] = {}
             dutPortIps[src_dut_index][src_asic_index] = {}
+            sysPortMap[src_dut_index][src_asic_index] = {}
             active_ips = src_asic.get_active_ip_interfaces(tbinfo)
             for iface, addr in active_ips.items():
                 if iface.startswith("Ethernet") and ("Ethernet-Rec" not in iface):
                     portIndex = src_mgFacts["minigraph_ptf_indices"][iface]
                     portIpMap = {'peer_addr': addr["peer_ipv4"], 'port': iface}
                     dutPortIps[src_dut_index][src_asic_index].update({portIndex: portIpMap})
+                    # Map port IDs to system port IDs
```
Comment (vmittal-msft) [inline tests/qos/qos_sai_base.py:972]: Same as above

File: tests/qos/qos_sai_base.py [hunk 3]
```diff
@@ -934,24 +954,43 @@ def dutConfig(
...
+                        sysPortMap[src_dut_index][src_asic_index].update({portIndex: sysPort})
+
                 elif iface.startswith("PortChannel"):
                     portName = next(
                         iter(src_mgFacts["minigraph_portchannels"][iface]["members"])
                     )
                     portIndex = src_mgFacts["minigraph_ptf_indices"][portName]
                     portIpMap = {'peer_addr': addr["peer_ipv4"], 'port': portName}
                     dutPortIps[src_dut_index][src_asic_index].update({portIndex: portIpMap})
+                    # Map lag port IDs to system port IDs
```
Comment (vmittal-msft) [inline tests/qos/qos_sai_base.py:986]: Same as above

File: tests/saitests/py3/sai_qos_tests.py [hunk 4]
```diff
@@ -3704,6 +3720,13 @@ def runTest(self):
                 for cntr in egress_counters:
                     assert (xmit_counters[cntr] > xmit_counters_base[cntr])
 
+            # voq ingress drop
+            if platform_asic and platform_asic == "broadcom-dnx":
+                voq_index = pg - 2
```
Comment (vmittal-msft) [inline tests/saitests/py3/sai_qos_tests.py:3725]: Please share more info on this logic

File: tests/saitests/py3/switch.py [hunk 5]
```diff
@@ -822,6 +822,33 @@ def sai_thrift_read_port_counters(client, asic_type, port):
...
+
+
+def sai_thrift_read_port_voq_counters(client, voq_list):
+    cnt_ids = []
+    thrift_results = []
+    voq_counters_results = []
+    cnt_ids.append(SAI_QUEUE_STAT_PACKETS)
+    queue1 = 0
+    for voq in voq_list:
+        if queue1 <= 7:
```
Comment (vmittal-msft) [inline tests/saitests/py3/switch.py:844]: Why do we need queue1 here ? Can't we work with VoQ id directly ?

File: tests/qos/qos_sai_base.py (hunk 1 above)
Comment (ansrajpu-git) [reply tests/qos/qos_sai_base.py]: done

File: tests/qos/qos_sai_base.py (hunk 2 above)
Comment (ansrajpu-git) [reply tests/qos/qos_sai_base.py]: done

File: tests/qos/qos_sai_base.py (hunk 3 above)
Comment (ansrajpu-git) [reply tests/qos/qos_sai_base.py]: done

File: tests/saitests/py3/sai_qos_tests.py (hunk 4 above)
Comment (ansrajpu-git) [reply tests/saitests/py3/sai_qos_tests.py:3725]: As we discussed, the pg value is set by adding 2 (pg = int(self.test_params['pg']) + 2) in the  initialization. Hence  reducing  2 here to get the correct voq_index

File: tests/saitests/py3/switch.py (hunk 5 above)
Comment (ansrajpu-git) [reply tests/saitests/py3/switch.py]: Changed the naming convention to 'counter'. It is just a counter value.

**Comment by arlakshm** [comment]
@kenneth-arista and @ysmanman for viz..

**Review by vmittal-msft — APPROVED** [review]

**Comment by vmittal-msft** [comment]
@ansrajpu-git please fix conflicts.
@yxieca @wangxin please help merge.

**Comment by ansrajpu-git** [comment]
> @ansrajpu-git please fix conflicts. @yxieca @wangxin please help merge.

@vmittal-msft ,resolved conflicts.

**Comment by vmittal-msft** [comment]
@yxieca @wangxin please help merge

**Comment by ansrajpu-git** [comment]
Decided to postpone this PR merge to 202205 branch now, as this introduces new SAI modules.

**Comment by XuChen-MSFT** [comment]
@ansrajpu-git  @vmittal-msft  need to cherry-pick to 202311 ?

+ @StormLiangMS  for viz....

**Comment by Janetxxx** [comment]
@yxieca Could you please help to approve the label? Thanks! ```sai_qos_tests.py``` is trying to import ```sai_thrift_read_port_voq_counters``` from the ```switch.py```, hence this PR needs to be included into 202311.

**Pull request merged**